### DISCUSSION TITLE
Permit implicit casting operators for pointers

### DISCUSSION THREAD

**Discussion by rickbrew** [dotnet/csharplang#5160] [Language Ideas]
We've been talking about this over on the C# Discord server in `#lowlevel`, and permitting casting operators for `T*` would be very useful for interop code. The specific scenario is to enable better pointer casts for something like TerraFX.Interop.Windows https://github.com/terrafx/terrafx.interop.windows

In C, C++, and in C++/CLI, you have no trouble doing the following:

```
int CreateBitmapBrush(ID2D1Brush* pBrush, out ID2D1BitmapBrush* pBitmapBrush) { ... }

ID2D1SolidColorBrush *pBrush = ... create a solid color brush ...
int hresult = CreateBitmapBrush(pBrush, out ID2D1BitmapBrush* pBitmapBrush);
```

In this case, the native `ID2D1SolidColorBrush` inherits from `ID2D1Brush` (which inherits from `ID2D1Resource`, then `IUnknown`). C and C++ handle the downcast natively, and the C++/CLI compiler knows about the inheritance and inserts the cast in the generated IL (which you can see in a disassembly).

In C++/CLI you can just `#include <d2d1.h>` and it all works. In C# you reference the `TerraFX.Interop.Windows` package which defines structs for each of the interfaces, e.g.

```
[Guid("00000000-0000-0000-C000-000000000046")]
public struct IUnknown
{
    public unsafe void** lpVtbl;
    ...
}
```

... but when it comes time to cast pointers, you have to do the cast manually and the compiler can't verify its safety. There's nothing stopping you from casting, say, `IWICBitmap*` to `ID2D1Bitmap*`. The compiler is happy, since explicit pointer casts are always valid. You can use `QueryInterface` but that is costly, and still lacks compile-time safety checks (albeit with a runtime error, which is preferably to, say, calling the wrong vtable methods and/or AV'ing).

But you also can't do this, which would make this work:

```
public struct ID2D1SolidColorBrush
{
    public unsafe void** lpVtbl;
    ...
    public static unsafe implicit operator ID2D1Brush*(ID2D1SolidColorBrush *p) { return (ID2D1Brush*)p; }
    ...
}
```

because of CS0556, "User-defined conversion must convert to or from the enclosing type." This is understandable in the general case because otherwise the compiler / runtime would have a _much_ larger space to search through for casting operators and I'm sure performance would be terrible (not to mention there could be ambiguity problems, etc.)

But this is a special case for `T* -> U*` and probably `T** -> U**`, so the search space is tightly constrained. Being able to have these operators, and therefore being able to codegen all of them, would greatly improve the safety of down-to-the-metal interop code like what TerraFX affords.

I'm currently researching and planning for how I'm going to go about converting a large chunk of C++/CLI code over to C#. This code is in Paint.NET's custom interop layer for Direct2D, WIC, UIAnimation, et. al. I'm planning to use TerraFX, and this language feature would make things so much easier and safer.

cc @tannergooding

**Comment by Sergio0694** [comment]
+1 to this. Not being able to have any kind of type safety at all when dealing with pointer casts is pretty inconvenient and makes working with lowlevel bindings more dangerous than necessary (and they're already not the most user friendly experience 😄). Allowing library authors to define pointer casts like this would seem like a nice way to fix this scenario without having to make major changes to the language and type system.

I'm using `TerraFX.Interop.Windows` extensively in [`ComputeSharp`](https://github.com/Sergio0694/ComputeSharp) and something like this would definitely make my life easier 👍

**Comment by svick** [comment]
If this is more about safety than about brevity, wouldn't a regular static method (possibly combined with `using static`) work almost as well?

E.g. instead of the desired, but currently not working `CreateBitmapBrush(pBrush, ...)` or the unsafe `CreateBitmapBrush((ID2D1Brush*)pBrush, ...)`, you would write something like  `CreateBitmapBrush(AsID2D1Brush(pBrush), ...)`.

(An extension method would be even better, but those are also not allowed on pointers.)

**Reply by Sergio0694** [reply]
> "An extension method would be even better, but those are also not allowed on pointers"

A bit off-topic, but just as a side note, extension methods on pointers do actually exist in C# 😄
They're just a bit unintuitive to setup, and they're not really documented anywhere that I could see.

This works just fine though, for instance:

```csharp
int x = 10;
int* p = &x;
    
p->InterlockedAdd(42); // Ta-dà! 🎉

public static class Extensions
{
    public static int InterlockedAdd(this ref int x, int y)
    {
        return Interlocked.Add(ref x, y);
    }
}
```

Basically you just need to define a `this ref` extension method, and then the compiler will happily pick that up as extension method for a pointer when using the `->` operator on it, and do the conversion to `ref` for you automatically to make it work 🙂

**Comment by phizch** [comment]
The really big issue with working with COM is that inheritance plays a big part since the interfaces can't really be changed after they're published. Here's on such inheritance chain.
```c++
struct ID3D12Device9 : public ID3D12Device8
struct ID3D12Device8 : public ID3D12Device7
struct ID3D12Device7 : public ID3D12Device6
struct ID3D12Device6 : public ID3D12Device5
struct ID3D12Device5 : public ID3D12Device4
struct ID3D12Device4 : public ID3D12Device3
struct ID3D12Device3 : public ID3D12Device2
struct ID3D12Device2 : public ID3D12Device1
struct ID3D12Device1 : public ID3D12Device
struct ID3D12Device : public ID3D12Object
struct ID3D12Object : public IUnknown
```
Even with implicit cast, it's 66 cast operators in total, and that's just for that chain.

`ID3D12Object` has a method `SetName(  ushort* Name )`. In c# this is unwieldy to use:
```cs
ID3D12Device* _device;
...
fixed( char* pName = "MyDevice" )
{
    _device->SetName( (ushort*)pName );
}
``` 
A solution would be to create an extension
```cs
  public static void SetName( ref this ID3D12Device self, string name )
  {
    fixed( char* pName = name )
    {
      self.SetName( (ushort*)pName );
    }
  }
...
ID3D12Device* _device;
...
_device->SetName( "MyDevice" );
```
This works, but unfortunately, we have to create one extension for each of the other versions of the interface. Also for *all* the other interfaces that's derived from `ID3DObject`. Also, there are other methods on `D3D12Object`, and it's descendants, that could use helper methods too.

With the ability to create implicit cast operators for pointers, we wouldn't gain the ability to create just one extension method¹, but it opens up the following:
```cs
static class Helpers
{
  public static void SetName( ID3D12Object* obj, string name ) {...}
}
...
using static Helpers;
...
ID3D12Device* _device;
...
SetName( _device, "MyDevice" ); // no compiler error, _device is implicitly cast to ID3D12Object*
```

Therein lies the safety bit. 

Today it is possible to have a `SetName` method that takes `( void* whatever )`. It will work, but unless the whatever actually *is* a `ID3D12Object` or derived, _bad stuff happens_.
```cs
int MyInt = 5;
int* pMyInt = &MyInt;

// with SetName( void*, string )
SetName( pMyInt, "MyDevice" ); // no compile error, and it doesn't look wrong at first glance.

// with SetName( ID3D12Object*, string )
SetName( (ID3D12Object*)pMyInt, "MyDevice" ); // no compiler error, but at least you have a chance to notice it.
SetName( pMyInt, "MyDevice" ); // compiler error
```

It is also possible to explicitly cast *any* pointer to any other pointer. That ability can't be removed, but with the added ability to create *implicit* casts, some "safety" in `unsafe` can be regained.

¹: It could, if at the same time some mechanism was put in place that could tell the compiler that the structs are "interited". That way extension methods could be created on the base version and all sub-"classes" could use the same extension method, just like with C# classes and interfaces.

**Comment by BinToss** [comment]
Implicit return-pointer operators seem to have been implemented at some point.
The following is valid in a .NET 6 C#10  project
```cs
    public struct PH_CONDITION
    {
        private PH_QUEUED_LOCK _;
        public nuint Value { get => _.Value; set => _.Value = value; }
        public static implicit operator PH_CONDITION(PH_QUEUED_LOCK value) => new() { _ = value };
        public static implicit operator PH_QUEUED_LOCK(PH_CONDITION value) => value._;
    }

    public struct PPH_CONDITION
    {
        private PH_CONDITION* _;
        public static implicit operator PH_CONDITION*(PPH_CONDITION value) => value._;
        public static implicit operator PPH_CONDITION(PH_CONDITION* value) => new() { _ = value };
    }
```

However, implicit conversion to a pointer type is invalid in another project that targeted NETFX 4.6.2 up to .NET 6.

On an unrelated note, when porting c/c++ to c#, it's faster to create a codegen pattern to create pointer-type wrappers when an incomplete type definition's symbol matches regex:`^P(\w)*`.

**Reply by svick** [reply]
As far as I can tell, operators converting from `T` to `U*` and from `U*` to `T` were always allowed (as long as they were defined on `T`). What this discussion is about are operators converting from `T*` to `U*`.

**Reply by BinToss** [reply]
Ah, I see now. That would be pretty useful.

**Comment by rickbrew** [comment]
@phizch, there have been some updates in this space for TerraFX and pointer casting.

The first is that the COM structs in TerraFX now use interfaces:

```cs
public struct IUnknown : IUnknown.Interface
{
    public interface Interface { ... }
}

public struct ID3D12Object : ID3D12Object.Interface
{
    public interface Interface : IUnknown.Interface { ... }
}

...

public struct ID3D12Device9 : ID3D12Device9.Interface
{
    public interface Interface : ID3D12Device8.Interface { ... }
}
```

This makes it possible to write extension methods that work on all of them, like so:
```cs
public static HRESULT SetName<T>(this ref T pD3D12Device, string name)
    where T : unmanaged, ID3D12Device.Interface
{
    fixed (char* pName = name)
    {
        return pD3D12Device.SetName((ushort*)pName);
    }
}
```

In addition, I've published two nuget packages, https://github.com/rickbrew/PointerToolkit and https://github.com/rickbrew/PointerToolkit.TerraFX.Interop.Windows. The first defines, among other things, a series of `CastPtr<T, TBase1, ... TBaseN>` types, and the latter contains generated `__cast(IFoo* p)` methods, one for each COM struct defined in TerraFX, e.g.

```cs
[MethodImpl(MethodImplOptions.AggressiveInlining)]
public static CastPtr<
    TerraFX.Interop.DirectX.ID3D12Device9, 
    TerraFX.Interop.DirectX.ID3D12Device8, 
    TerraFX.Interop.DirectX.ID3D12Device7, 
    TerraFX.Interop.DirectX.ID3D12Device6, 
    TerraFX.Interop.DirectX.ID3D12Device5, 
    TerraFX.Interop.DirectX.ID3D12Device4, 
    TerraFX.Interop.DirectX.ID3D12Device3, 
    TerraFX.Interop.DirectX.ID3D12Device2, 
    TerraFX.Interop.DirectX.ID3D12Device1, 
    TerraFX.Interop.DirectX.ID3D12Device, 
    TerraFX.Interop.DirectX.ID3D12Object, 
    TerraFX.Interop.Windows.IUnknown> __cast(TerraFX.Interop.DirectX.ID3D12Device9* ptr)
{
    return *(CastPtr<TerraFX.Interop.DirectX.ID3D12Device9, TerraFX.Interop.DirectX.ID3D12Device8, TerraFX.Interop.DirectX.ID3D12Device7, TerraFX.Interop.DirectX.ID3D12Device6, TerraFX.Interop.DirectX.ID3D12Device5, TerraFX.Interop.DirectX.ID3D12Device4, TerraFX.Interop.DirectX.ID3D12Device3, TerraFX.Interop.DirectX.ID3D12Device2, TerraFX.Interop.DirectX.ID3D12Device1, TerraFX.Interop.DirectX.ID3D12Device, TerraFX.Interop.DirectX.ID3D12Object, TerraFX.Interop.Windows.IUnknown>*)&ptr;
}
```

Using generics greatly reduces the code size. Instead of one casting helper method per combination of valid pointer casts (which is a lot), you only need 1 helper method per interface. I tried the first method and the DLL was 3MB, IIRC, and really slowed down VS and Intellisense, but now it's only about 300KB and performance is good.

To use this, you add a `global using static TerraFX.Interop.DirectX.Pointers;` to your project, and then you can use `__cast(p)` where `p` is e.g. an `ID3D12Device9*`. You then get a `ref struct CastPtr<ID3D12Device9, ..., IUnknown>` with implicit cast operators to all of the base struct types.

Then you can do things like:

```cs
public static HRESULT SetName(ID3D12Device* pDevice, string name)
{
    ...
}

ID3D12Device9* pDevice = ...;
SetName(__cast(pDevice), "Name");
```

This fixes the Direct2D example I mention at the top, just do `__cast(pBrush)` when calling a method that takes an `ID2D1Brush*` when `pBrush` is an `ID2D1SolidColorBrush*`.

It's not as elegant as having real implicit pointer casts, but it's good 'nuff and meets the compile-time safety requirements I was grasping for last September.
